src/Request.ts

```diff
@@ -0,0 +1,350 @@
+		 * parsing function. If you do want to use an alternate format, you will probably
+		 * want to use an alternate stringify function for the serialization of data as well.
+		 * Also, if you want to support parsing a larger set of JavaScript objects
+		 * outside of strict JSON parsing, you can provide dojo/_base/json.fromJson as the parse function
+		 */
+		this.parse = this.parse || function (data: string) {
+				return JSON.parse(data);
+			};
+		this.headers = this.headers || {};
+		this.target = this.target || '';
+		this._targetContainsQueryString = this.target.lastIndexOf('?') >= 0;
+		this.ascendingPrefix = this.ascendingPrefix || '+';
+		this.descendingPrefix = this.descendingPrefix || '-';
+		this.accepts = this.accepts || 'application/json';
+		this.fetch = this.fetch || function (kwArgs?: dstore.FetchArgs) {
+			const results = this._request(kwArgs);
+			return <dstore.FetchPromise<any>> QueryResults(results.data, {
+				response: results.response
+			});
+		};
+	}
+
+	protected _request(kwArgs: dstore.FetchArgs = {}): {
+		data: dstore.FetchPromise<T>;
+		total: Promise<number>;
+		response: ResponsePromise<T>
+	} {
+		// perform the actual query
+		const RequestHeader: { [ name: string ]: string } = null;
+		const headers = <typeof RequestHeader> lang.mixin(Object.create(this.headers), { Accept: this.accepts });
+
+		if ('headers' in kwArgs) {
+			lang.mixin(headers, kwArgs.headers);
+		}
+
+		const requestUrl = this._renderUrl(kwArgs.queryParams);
+
+		const response = request(requestUrl, {
+			method: 'GET',
+			headers: headers
+		});
+		const collection = this;
+		const parsedResponse = response.then(function (response: Response<string>) {
+			return collection.parse(response.data);
+		});
+		return {
+			data: parsedResponse.then(function (data) {
+				// support items in the results
+				const results = data.items || data;
+				for (let i = 0, l = results.length; i < l; i++) {
+					results[ i ] = collection._restore(results[ i ], true);
+				}
+				return results;
+			}),
+			total: parsedResponse.then(function (data) {
+				// check for a total property
+				const total = data.total;
+				if (total > -1) {
+					// if we have a valid positive number from the data,
+					// we can use that
+					return total;
+				}
+				// else use headers
+				return response.then(function (response) {
+					const rangeHeader = response.getHeader('Content-Range');
+					let rangeMatch: RegExpMatchArray;
+					let range: number;
+					if (rangeHeader && (rangeMatch = rangeHeader.match(/\/(.*)/)) && (range = Number(rangeMatch[ 1 ]))) {
+						return range;
+					} else {
+						return null;
+					}
+				});
+			}),
+			response: response
+		};
+	}
+
+	/**
+	 * Constructs filter-related params to be inserted into the query string
+	 *
+	 * @param filter The filter to render as part of a query string
+	 * @returns Filter-related params to be inserted in the query string
```

Comment:
`@return` for consistency with elsewhere (there are ~5 other occurrences of this in this changeset)